lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1028 @@
+
+		// AbstractSparseLattice
+		// A lattice contains information about an SSA value and is what's propagated across the IR by sparse data-flow analysis.
+
+		// AbstractDenseForwardDataFlowAnalysis
+		// Dense data-flow analysis attaches a lattice between the execution of operations and implements a transfer function from the lattice before each operation to the lattice after. The lattice contains information about the state of the program at that point.
+
+		// In this implementation, a lattice attached to an operation represents the state of the program after its execution, and a lattice attached to block represents the state of the program right before it starts executing its body. 
+
+		// AbstractSparseForwardDataFlowAnalysis
+		// A sparse analysis implements a transfer function on operations from the lattices of the operands to the lattices of the results. This analysis will propagate lattices across control-flow edges and the callgraph using liveness information. 
+
+		// My initial conclusion is then that we need an AbstractDenseLattice and an AbstractDenseBackwardDataFLowAnalysis
+
+		// So our lattice is attached to operations -> for each operation we need to store information about the operands and results
+		// This means that we need a set (more precisely a dictionary) since we do not know how many operands need to be saved
+
+		//TODO: Can i make it generic wrt an integer type?
+		class ConstraintsLattice: public mlir::dataflow::AbstractDenseLattice
+		{
+
+			public:
+
+			using Integer=mlir::rlc::IntegerRange::Integer;
+			constexpr static Integer MAX=INT32_MAX;
+			constexpr static Integer MIN=INT32_MIN;
+
+			MLIR_DEFINE_EXPLICIT_INTERNAL_INLINE_TYPE_ID(ConstraintsLattice);
+			explicit ConstraintsLattice(mlir::ProgramPoint point)
+					: mlir::dataflow::AbstractDenseLattice(point)
+			{
+			}
+
+			// Method for returning the lattice (useful only at the end) <- maybe can do everything in the class
+			const llvm::DenseMap<std::pair<mlir::Value,bool>,IntegerRange>& getUnderlyingLattice() const{
+				return this->ranges;
+			}
+
+			//TODO: understand what this meet does
+			// In theory it should be the opposite of a join
+			mlir::ChangeResult meet(
+					const mlir::dataflow::AbstractDenseLattice& val) override
+			{
+				llvm::outs()<<"meet\n";
+				//return mlir::ChangeResult::NoChange;
+				// Do a funny thing and let the meet perform the join
+				return this->join(val);
+			}
+
+			mlir::ChangeResult join(
+					const mlir::dataflow::AbstractDenseLattice& val) override
+			{
+				//Need to cast it
+				const auto& other = *static_cast<const ConstraintsLattice*>(&val);
+				
+				//Keep it to check if there is no change
+				auto copy = *this;
+
+				llvm::outs()<<"Did i join?\n";
+				this->print(llvm::outs());
+				other.print(llvm::outs());
+
+				//Insert the other in this
+				for(auto other_range : other.ranges){
```

Comment:
const auto& instead of auto, no reason to make a copy here!

Reply:
hm I did not know that it would make a copy. Very interesting

Reply:
yeah, it is a cpp pitfall, auto means the type that is on the right, without references. so you can discriminate between copies and non copies with `auto` and `auto&`